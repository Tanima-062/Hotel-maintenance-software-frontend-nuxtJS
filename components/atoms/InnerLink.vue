<template>
  <nuxt-link
    class="cursor-pointer focus:outline-none"
    :class="styles"
    :to="href"
  >
    <slot />
  </nuxt-link>
</template>
<script lang="ts">
import { Vue, Component, Prop } from 'nuxt-property-decorator'

@Component
export default class extends Vue {
  @Prop({ default: '' })
  href!: string

  @Prop({ default: 'default' })
  textColor!: string

  @Prop({ default: 'default' })
  hoverTextColor!: string

  @Prop({ default: true })
  showUnderline!: boolean

  @Prop({ default: true })
  focusOn!: boolean

  get styles(): string {
    const classList = []
    if (this.showUnderline) {
      classList.push('underline')
    }
    classList.push(`${this.textColor}`)
    classList.push(`hover:${this.hoverTextColor}`)
    classList.push(this.focusOn ? 'focus:shadow-outline' : 'focus:shadow-none')
    return classList.join(' ')
  }
}
</script>
