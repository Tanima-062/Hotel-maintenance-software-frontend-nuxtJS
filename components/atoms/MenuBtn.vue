<template>
  <button
    type="button"
    class="flex justify-center items-center leading-5 whitespace-no-wrap"
    :class="styleColor"
    @click="click"
    @mouseover="mouseover"
    @mouseleave="mouseleave"
  >
    <slot />
  </button>
</template>
<script lang="ts">
import { Vue, Component, Prop, Emit } from 'nuxt-property-decorator'

@Component
export default class extends Vue {
  @Prop({ default: 'text-black' })
  textColor!: string

  @Prop({ default: false })
  isActive!: boolean

  private isAnim: boolean = false
  private interval: number | undefined

  get styleColor(): string {
    const classList = []
    // click時のボタン
    classList.push(this.isPushed ? 'btn--click' : 'btn')

    // テキスト文字の切り替え
    classList.push(this.textColor)

    // 背景色の切り替え
    classList.push(this.isPushed ? 'bg-menuColor-hover font-bold' : '')

    return classList.join(' ')
  }

  get isPushed(): boolean {
    return this.isAnim || this.isActive
  }

  private doAnim(): void {
    if (this.isAnim || this.interval) {
      return
    }
    this.isAnim = true
    this.interval = window.setInterval(() => {
      this.isAnim = false
      clearInterval(this.interval)
      this.interval = undefined
    }, 150)
  }

  @Emit()
  private click(e: Event): Event {
    this.doAnim()
    return e
  }

  @Emit()
  mouseover(e: Event): Event {
    return e
  }

  @Emit()
  mouseleave(e: Event): Event {
    return e
  }
}
</script>
<style lang="sass" scoped>
.btn
  transition: background-color .3s ease-in
  &:hover
    transition: opacity .1s cubic-bezier(.4,0,.6,1)
    opacity:.9
  &--click
    transition: background-color .15s ease
</style>
