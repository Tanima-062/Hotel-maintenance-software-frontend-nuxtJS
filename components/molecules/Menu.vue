<template>
  <div class="w-full shadow h-11 bg-menuColor flex items-center justify-center">
    <transition-group
      class="w-full max-w-1200 flex items-center z-30"
      name="menu-drop"
      tag="div"
    >
      <div
        v-for="item in menuTabs"
        :key="item.key"
        tag="div"
        class="relative"
        @mouseleave="setHoverBtnKey('')"
      >
        <MenuBtn
          class="py-3"
          :value="item.key"
          :class="activeStyle(item.key)"
          :is-active="isActive(item.key)"
          color="text-menu"
          @mouseover="setHoverBtnKey(item.key)"
          @click="onClick"
        >
          {{ item.name }}
        </MenuBtn>
        <div
          v-if="isHoverTabKey === item.key"
          class="absolute origin-top mt-3px"
          :class="btnWidth"
        >
          <ul
            class="border border-t-0 border-bdColor border-solid bg-menuColor w-full menuItem-area text-center"
          >
            <li
              v-for="(liItem, liItemIndex) in item.items"
              :key="liItemIndex"
              class="p-3 text-sm hover:text-menuColor-active"
              :class="`
                ${liItemIndex > 0 && 'border-t border-bdColor border-solid'} ${
                liItem.href === $route.path &&
                'text-menuColor-active font-normal'
              }
              `"
            >
              <InnerLink
                :href="liItem.href"
                :show-underline="false"
                hover-text-color="text-menuColor-active"
                :focus-on="false"
              >
                {{ liItem.name }}
              </InnerLink>
            </li>
          </ul>
        </div>
      </div>
    </transition-group>
  </div>
</template>
<script lang="ts">
import { Vue, Component, Prop, Emit } from 'nuxt-property-decorator'

@Component({
  components: {
    MenuBtn: () => import('~/components/atoms/MenuBtn.vue'),
    InnerLink: () => import('~/components/atoms/InnerLink.vue'),
  },
})
export default class extends Vue {
  @Prop({ default: [], required: true })
  menuTabs!: Object[]

  @Prop({ default: 'w-40' })
  btnWidth!: string

  @Prop({ default: '' })
  activeTabKey!: string

  private isHoverTabKey: string = ''

  private focusTabItemStyle(liItemIndex: number): string {
    const classList = []
    if (!liItemIndex) {
      classList.push('border-t border-bdColor border-solid')
    }
    classList.push(this.btnWidth)
    return classList.join(' ')
  }

  private activeStyle(tabKey: string): string {
    const classList = []
    classList.push(this.btnWidth)
    if (this.isActive(tabKey)) {
      classList.push('border-b-3 border-solid border-menuColor-active -mb-3px')
    }
    return classList.join(' ')
  }

  private isActive(tabKey: string): boolean {
    return this.activeTabKey === tabKey || this.isHoverTabKey === tabKey
  }

  private setHoverBtnKey(v: string): void {
    this.isHoverTabKey = v
  }

  @Emit()
  private onClick(e: Event & { target: HTMLInputElement }): string {
    return e.target.value
  }
}
</script>
<style lang="sass" scoped>
.anim
  transition: all .3s ease

.menu-drop-enter-active, .menu-drop-leave-active
  transition: opacity .5s

.menu-drop-enter, .menu-drop-leave-to
  opacity: 0

.menuItem-area
  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.06)
</style>
