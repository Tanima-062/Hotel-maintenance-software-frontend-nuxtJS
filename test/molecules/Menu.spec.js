import { mount } from '@vue/test-utils'
import Menu from '@/components/molecules/Menu.vue'
import MenuBtn from '@/components/atoms/MenuBtn.vue'
import InnerLink from '@/components/atoms/InnerLink.vue'

describe('Molecules Menu Check', () => {
  test('Menu Test', () => {
    const props = {
      menuTabs: [
        {
          key: 'booking',
          name: 'booking',
          items: [
            {
              href: '/booking',
              name: 'booking',
              isShowOnMenu: true,
            },
          ],
        },
      ],
    }
    const wrapper = mount(Menu, {
      propsData: props,
      stubs: {
        MenuBtn,
        InnerLink,
      },
    })
    // check if it is a vue element
    expect(wrapper.vm).toBeTruthy()

    // check Snapshot
    expect(wrapper.element).toMatchSnapshot()
  })
})
