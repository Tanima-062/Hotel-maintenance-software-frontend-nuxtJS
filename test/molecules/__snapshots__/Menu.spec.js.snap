// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecules Menu Check Menu Test 1`] = `
<div
  class="w-full shadow h-11 bg-menuColor flex items-center justify-center"
>
  <transition-group-stub
    class="w-full max-w-1200 flex items-center z-30"
    name="menu-drop"
    tag="div"
  >
    <div
      class="relative"
      tag="div"
    >
      <button
        class="flex justify-center items-center leading-5 whitespace-no-wrap py-3 btn text-black  w-40"
        color="text-menu"
        type="button"
        value="booking"
      >
        
        booking
      
      </button>
       
      <!---->
    </div>
  </transition-group-stub>
</div>
`;
